src/vm/functions/sequences.rs

```diff
@@ -0,0 +1,177 @@
+use vm::costs::{cost_functions, CostOverflowingMath};
+use vm::errors::{CheckErrors, RuntimeErrorType, InterpreterResult as Result, check_argument_count};
+use vm::types::{Value, SequenceData, CharType, ListData, signatures::ListTypeData, TypeSignature::BoolType, TypeSignature};
+use vm::representations::{SymbolicExpression, SymbolicExpressionType};
+use vm::{LocalContext, Environment, CallableType, eval, apply, lookup_function};
+use std::convert::TryInto;
+use std::cmp;
+
+pub fn list_cons(args: &[SymbolicExpression], env: &mut Environment, context: &LocalContext) -> Result<Value> {
+    let eval_tried: Result<Vec<Value>> =
+        args.iter().map(|x| eval(x, env, context)).collect();
+    let args = eval_tried?;
+
+    let mut arg_size = 0;
+    for a in args.iter() {
+        arg_size = arg_size.cost_overflow_add(a.size().into())?;
+    }
+
+    runtime_cost!(cost_functions::LIST_CONS, env, arg_size)?;
+
+    Value::list_from(args)
+}
+
+pub fn special_filter(args: &[SymbolicExpression], env: &mut Environment, context: &LocalContext) -> Result<Value> {
+    check_argument_count(2, args)?;
+
+    runtime_cost!(cost_functions::FILTER, env, 0)?;
+
+    let function_name = args[0].match_atom()
+        .ok_or(CheckErrors::ExpectedName)?;
+
+    let mut sequence = eval(&args[1], env, context)?;
+    let function = lookup_function(&function_name, env)?;
+
+    match sequence {
+        Value::Sequence(ref mut sequence_data) => {
+            sequence_data.filter(&mut |atom_value: SymbolicExpression| {
+                let argument = [atom_value];
+                let filter_eval = apply(&function, &argument, env, context)?;
+                if let Value::Bool(include) = filter_eval {
+                    return Ok(include);
+                } else {
+                    return Err(CheckErrors::TypeValueError(BoolType, filter_eval).into())
+                }
+            })?;
+        },
+        _ => return Err(CheckErrors::ExpectedSequence(TypeSignature::type_of(&sequence)).into())
+    };
+    Ok(sequence)
+}
+
+pub fn special_fold(args: &[SymbolicExpression], env: &mut Environment, context: &LocalContext) -> Result<Value> {
```

Comment:
Can `special_fold` be implemented with fewer copies? Currently, it seems like `.atom_values()` will make 1 copy of the input, and then `try_fold` copies again. It seems like `.atom_values()` could be a destructive function, which would eliminate that copy, and then try_fold could be called on `into_iter()` which would eliminate the second copy as well.

Reply:
Addressed with https://github.com/blockstack/stacks-blockchain/pull/1779/commits/27ff63ebd81f8d5566b87c04cbade46db81a2f51